server/serverAPI/controller/AuthenticationController.ts

```diff
@@ -101,6 +118,93 @@ export default class AuthenticationController extends BaseUserController {
         }
     }
 
+    confirmVerificationCode = async (req: Request, res: Response) => {
+        let username = req.body?.username;
+        let inputCode = req.body?.verificationCode;
+
+        let actualCode = AuthenticationController.verificationCodesMap.get(username);
+
+        try {
+            await this.requestGet(new Map([["username", username]]), res)
+        } catch (response) {
+            return response;
+        }
+
+        if (actualCode === undefined) {
+            return this.sendError(400, res, "Verification code is either expired or not issued.");
+        }
+
+        if (inputCode !== actualCode) {
+            return this.sendError(400, res, "Verification code is invalid.");
+        }
+
+        try {
+            let user = await this.requestGet(new Map([["username", username]]), res);
+
+            user.isVerified = true;
+
+            await this.requestUpdate(user.username, user, res);
+
+            return this.sendSuccess(200, res, "Account has been verified.");
+        } catch (e) {
+            return e;
+        }
+    }
+
+    sendVerificationCode = async (req: Request, res: Response) => {
+        let email = "";
+        let username = req.body?.username;
+
+        try {
+            let user = await this.requestGet(new Map([["username", username]]), res);
+            email = user.email;
+        } catch (e) {
+            return e;
+        }
+
+        if (AuthenticationController.verificationCodesMap.has(username)) {
+            let generationTime = AuthenticationController.verificationCodesMap.get(username)!.geneationTime;
+            let currentTime = Date.now();
+
+            if ((currentTime - generationTime) < this.verificationCodeLifetimeInMilliseconds) {
+                return this.sendError(400, res, "The previous code sent hasn't expired yet.");
+            }
```

Comment:
Won't the code be deleted from the map after the generationTime has passed? Is this logic needed? Could you just check to see if the map has the username?

Reply:
u are right...